### ISSUE TITLE
eval-table issue (MOSES)

### ISSUE THREAD

**Issue by edyirdaw** [opencog/opencog#1150]
Assume we have output of combo programs from a feature vector file
(training set) in sparse format. When I tried to use eval-table with a
new feature vector file (test set) in sparse format with one of the
combo programs, I get the following error below.
What happened is that, the variable $s0 which is found in the combo
program was not found in the test set. How do we tackle this?

Previously I tackled this issue by appending the training set to the
test set and the error would not occur. But this approach would be
very clumsy in my current situation. Is there a simpler way to solve
this problem?

Error:

opencog@iCog-Lab-4:~/eddy/misc2$ eval-table -i sparse2-test -c 'and(!$f0 !$s0)' -L1 -utar
terminate called after throwing an instance of 'opencog::AssertionException'
  what():  No label s0 matching
(/home/opencog/eddy/opencog/src/opencog/comboreduct/combo/iostream_combo.cc:455)
Aborted (core dumped)

Training file is sparse2 given below

tar
1,f1 : 1,s1 : 1
0,f0 : 1,s0 : 1
0,f0 : 1,s2 : 1
1,s1 : 1,s2 : 1

Test set is spare2-test given below

tar
1,f1 : 1,s1 : 1
0,f0 : 1,s5 : 1
0,f0 : 1,s2 : 1
1,s1 : 1,s2 : 1

**Comment by edyirdaw** [comment]
Hi @ngeiswei  and @linas , any progress on this issue yet? Thanks.

**Comment by linas** [comment]
't' is not a valid column name.  Its stupid to waste time to try to 'fix' this.  I'm not sure what issues are left, here, if any ...

**Comment by linas** [comment]
I don't understand what the problem actually is.   The sparse table does have defaults for the missing values -- that's the whole point of being sparse!  The default is zero-false, and there is no way to over-ride that value.

Do you have a test case that reproduces the problem?

**Comment by linas** [comment]
Oh wait .. never mind. I see that you did post this. Nil is right, sparse-test does not actually have s0 appearing in it, anywhere, at all. 

You have two choices: if there is a variable in the combo that is not in the test set, then throw an error. I guess that is what it currently does.

The other choice is to add that variable to the dataset, and give it the default value.

There is a third choice, but it is harder: remove the variable from the combo program... this last approach is probably wrong.  So the middle choice seems to be best: just add s0 to the test set (automatically)

**Comment by ngeiswei** [comment]
I see 2 options:

1. keep raising an error by default, add yet another user option to automatically add 0/false columns.
2. Add 0/false columns by default, but log a warning message when that happens.

Also, should this work too when the table is non spares?

I can take care of coding that, unless edyirdaw, you feel like doing it, let me know.

**Comment by edyirdaw** [comment]
Thanks for the replies @ngeiswei and @linas. Nil, it's better you take care of this code as I have no experience at all coding MOSES. Both of your options are good. I think option 2 is much better. But I don't see why raising a warning message is needed. If the variable in the combo is not in the test set, then by definition of the sparse format, it has a zero value. This is my opinion, I don't know from what scenarios you saw it; may be it would be useful to log warning messages in some application areas. May be adding a warning message might also be of interest to see which variables in the combo don't appear in the test set. So, final decision is yours.

When the table in non-sparse, my opinion is yea, it should work because if variable is absent it is supposed to mean it's value is zero (I can't think of exceptions for now). But if the programmer as a non-sparse format, then he is capable of building the non-sparse format and would handle this issue easily. May be its good to wait for somebody to raise this problem for non-sparse data before coding it to save effort.

**Comment by bgoertzel** [comment]
Thanks for your help with this Nil ... ;)

On Fri, Jan 16, 2015 at 4:32 AM, edyirdaw <notifications@github.com> wrote:

> Thanks for the replies @ngeiswei <https://github.com/ngeiswei> and @linas
> <https://github.com/linas>. Nil, it's better you take care of this code
> as I have no experience at all coding MOSES. Both of your options are good.
> I think option 2 is much better. But I don't see why raising a warning
> message is needed. If the variable in the combo is not in the test set,
> then by definition of the sparse format, it has a zero value. This is my
> opinion, I don't know from what scenarios you saw it; may be it would be
> useful to log warning messages in some application areas. May be adding a
> warning message might also be of interest to see which variables in the
> combo don't appear in the test set. So, final decision is yours.
>
> When the table in non-sparse, my opinion is yea, it should work because if
> variable is absent it is supposed to mean it's value is zero (I can't think
> of exceptions for now). But if the programmer as a non-sparse format, then
> he is capable of building the non-sparse format and would handle this issue
> easily. May be its good to wait for somebody to raise this problem for
> non-sparse data before coding it to save effort.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/opencog/opencog/issues/1150#issuecomment-70235514>.
>



-- 
Ben Goertzel, PhD
http://goertzel.org

"The reasonable man adapts himself to the world: the unreasonable one
persists in trying to adapt the world to himself. Therefore all progress
depends on the unreasonable man." -- George Bernard Shaw

**Comment by ngeiswei** [comment]
I've pushed a commit for that https://github.com/opencog/opencog/commit/8d1890b0c62fef17815a5449ba83a1845334dde4 . I don't log at WARN but at INFO level, which makes sense I think. It automatically infers the type of the missing features and use default values defined in function 
```
vertex default_vertex_value(type_node tn)
```
in file
```
opencog/comboreduct/type_checker/type_tree.cc
```
contin and boolean are supported and adding more is easy.

**Comment by linas** [comment]
Nil, it would be better if you did pull requests for this. That way, a change over multiple commits is automatically recorded in the log, and also, if you say things like "fixes issue #xyz" or "closes but #xyz" then the bug report is automatically closed.

**Comment by ngeiswei** [comment]
Good to know, thanks! (it's actually a single commit but yeah better create a branch per issue)

**Comment by edyirdaw** [comment]
Thanks Nil
